File: src/main/java/uk/edu/populationfitness/models/genes/fitness/FitnessRange.java
package uk.edu.populationfitness.models.genes.fitness;

public class FitnessRange {

    // Defines the minimum fitness expected from the fitness function
    private double min_fitness;

    // Defines the max fitness expected from the fitness function
    private double max_fitness;

    private double range;

    private Statistics statistics;

    public FitnessRange(Statistics statistics){
        min_fitness = 0;
        max_fitness = 1;
        range = 1;
        this.statistics = statistics;
    }

    public FitnessRange(){
        this(null);
    }

    public FitnessRange max(double max){
        max_fitness = max;
        range = max_fitness - min_fitness;
        return this;
    }

    public FitnessRange min(double min){
        min_fitness = min;
        range = max_fitness - min_fitness;
        return this;
    }

    public double max(){
        return max_fitness;
    }

    public double min(){
        return min_fitness;
    }

    public FitnessRange statistics(Statistics stats){
        this.statistics = stats;
        return this;
    }

    public Statistics statistics() {
        return statistics;
    }

    // returns the range as a scale. S(f) = (f - min) / (max - min)
    public double toScale(double fitness){
        if (statistics != null){
            statistics.add(fitness);
        }
        return (fitness - min_fitness) / range;
    }
}

File: DotNet/PopulationFitness/PopulationFitness/Models/Genes/Fitness/FitnessRange.cs
namespace PopulationFitness.Models.Genes.Fitness
{
    public class FitnessRange
    {

        // Defines the minimum fitness expected from the fitness function
        private double _minFitness;

        // Defines the max fitness expected from the fitness function
        private double _maxFitness;

        private double _range;

        private Statistics _statistics;

        private FitnessRange(Statistics statistics)
        {
            _minFitness = 0;
            _maxFitness = 1;
            _range = 1;
            this._statistics = statistics;
        }

        public FitnessRange() : this(null)
        {
        }

        public FitnessRange Max(double max)
        {
            _maxFitness = max;
            _range = _maxFitness - _minFitness;
            return this;
        }

        public FitnessRange Min(double min)
        {
            _minFitness = min;
            _range = _maxFitness - _minFitness;
            return this;
        }

        public double Max()
        {
            return _maxFitness;
        }

        public double Min()
        {
            return _minFitness;
        }

        public FitnessRange Statistics(Statistics stats)
        {
            _statistics = stats;
            return this;
        }

        public Statistics Statistics()
        {
            return _statistics;
        }

        // returns the range as a scale. S(f) = (f - min) / (max - min)
        public double ToScale(double fitness)
        {
            if (_statistics != null)
            {
                _statistics.Add(fitness);
            }
            return (fitness - _minFitness) / _range;
        }
    }
}
